internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/pubsub/api/consumer/PubSubConsumerAdapterTest.java

```diff
@@ -0,0 +1,1421 @@
+    assertTrue(pubSubConsumerAdapter.hasSubscription(partitionB0), "Should be subscribed to the topic and partition");
+    assertTrue(pubSubConsumerAdapter.hasSubscription(partitionB1), "Should be subscribed to the topic and partition");
+
+    // produce messages to the topic
+    PubSubProducerAdapter pubSubProducerAdapter = pubSubProducerAdapterLazy.get();
+    Map<PubSubTopicPartition, CompletableFuture<PubSubProduceResult>> lastMessageFutures = new HashMap<>(numPartitions);
+    for (int i = 0; i < numMessages; i++) {
+      lastMessageFutures.put(
+          partitionA0,
+          pubSubProducerAdapter
+              .sendMessage(topicA.getName(), 0, PubSubHelper.getDummyKey(), PubSubHelper.getDummyValue(), null, null));
+      lastMessageFutures.put(
+          partitionA1,
+          pubSubProducerAdapter
+              .sendMessage(topicA.getName(), 1, PubSubHelper.getDummyKey(), PubSubHelper.getDummyValue(), null, null));
+      lastMessageFutures.put(
+          partitionB0,
+          pubSubProducerAdapter
+              .sendMessage(topicB.getName(), 0, PubSubHelper.getDummyKey(), PubSubHelper.getDummyValue(), null, null));
+      lastMessageFutures.put(
+          partitionB1,
+          pubSubProducerAdapter
+              .sendMessage(topicB.getName(), 1, PubSubHelper.getDummyKey(), PubSubHelper.getDummyValue(), null, null));
+    }
+
+    CompletableFuture.allOf(lastMessageFutures.values().toArray(new CompletableFuture[0])).get(60, TimeUnit.SECONDS);
+    // check end offsets
+    long startTime = System.currentTimeMillis();
+    Map<PubSubTopicPartition, Long> endOffsets = pubSubConsumerAdapter.endOffsets(
+        new HashSet<>(Arrays.asList(partitionA0, partitionA1, partitionB0, partitionB1)),
+        PUBSUB_OP_TIMEOUT);
+    long elapsedTime = System.currentTimeMillis() - startTime;
+    assertTrue(elapsedTime <= PUBSUB_OP_TIMEOUT_WITH_VARIANCE, "endOffsets should not block");
+
+    assertEquals(endOffsets.get(partitionA0), Long.valueOf(numMessages), "End offset should match");
+    assertEquals(endOffsets.get(partitionA1), Long.valueOf(numMessages), "End offset should match");
+    assertEquals(endOffsets.get(partitionB0), Long.valueOf(numMessages), "End offset should match");
+    assertEquals(endOffsets.get(partitionB1), Long.valueOf(numMessages), "End offset should match");
+
+    Set<PubSubTopicPartition> consumptionBarMet = new HashSet<>();
+    while (consumptionBarMet.size() != 4) {
+      startTime = System.currentTimeMillis();
+      Map<PubSubTopicPartition, List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> messages =
+          pubSubConsumerAdapter.poll(1);
+      elapsedTime = System.currentTimeMillis() - startTime;
+      // check that poll did not block for longer than the timeout; add variance of 3 seconds
+      assertTrue(elapsedTime <= 1000 + 3000, "Poll should not block for longer than the timeout");
```

Comment:
Wondering how can we come up with time 1000 + 3000 ms, is it caused by default timeout settings?

Reply:
should be timeout passed to poll * retries + (retries - 1) * backoff. I can change it to some constant

Reply:
Yeah, that makes it easier to understand.

Reply:
Done